Sort OptionValues by position in filter metadata

aleksandarpetrushev opened the pull request.

Commit:
Sort OptionValues by position in filter metadata

Commit:
Sort OptionValues by position in filter metadata

Commit:
Sort OptionValues by position in filter metadata

Commit:
Sort OptionValues by position in filter metadata

damianlegawiec commented on core/app/presenters/spree/filters/options_presenter.rb:
```diff
@@ -9,7 +9,7 @@ def to_h
             name: option_type.name,
             presentation: option_type.presentation,
-            option_values: option_values.map { |e| option_value_hash(e) }
+            option_values: option_values.map { |e| option_value_hash(e) }.sort_by { |e| e[:position] }
```
@aleksandarpetrushev rather than doing this here - please fix it at the source - https://github.com/spree/spree/blob/a0137d7aa25fa48639c88e5baef11bd236559812/core/app/finders/concerns/spree/product_filterable.rb#L6

Commit:
Order option values by position in scope

Commit:
Order option values by position in scope

Commit:
Order option values by position in scope

Commit:
Order option values by position in scope

damianlegawiec commented on core/app/models/spree/option_value.rb:
```diff
@@ -16,6 +16,7 @@ class OptionValue < Spree::Base
     scope :filterable, lambda {
       joins(:option_type).
         where(OptionType.table_name => { filterable: true }).
+        order(position: :asc).
```
@aleksandarpetrushev 

1. this order is ambiguous as both option values and option types have the `position` attribute
2. Also let's not modify this scope, let's add the order in the service - https://github.com/spree/spree/blob/a0137d7aa25fa48639c88e5baef11bd236559812/core/app/finders/spree/option_values/find_available.rb

damianlegawiec replied:
3. We need tests to back it up

Commit:
Order option values in Finder service

Commit:
Order option values in Finder service

Commit:
Order option values in Finder service

Commit:
Order option values in Finder service

Commit:
Add specs for OptionValues Finder ordering

Commit:
Add specs for OptionValues Finder ordering

Commit:
Add specs for OptionValues Finder ordering

Commit:
Add specs for OptionValues Finder ordering

Commit:
Update OptionValue Finder spec

Commit:
Update OptionValue Finder spec

Commit:
Update OptionValue Finder spec

damianlegawiec commented on core/app/finders/spree/option_values/find_available.rb:
```diff
@@ -4,7 +4,7 @@ class FindAvailable
 
       def initialize(scope: OptionValue.spree_base_scopes, products_scope: Product.spree_base_scopes)
-        @scope = scope
+        @scope = scope.order(position: :asc)
```
@aleksandarpetrushev we need to sort this by option's type position first, and later by option value position

Commit:
Order OptionValues by OptionTypes position first

Commit:
Order OptionValues by OptionTypes position first

Commit:
Order OptionValues by OptionTypes position first

Commit:
Order OptionValues by OptionTypes position first

aleksandarpetrushev commented:
@damianlegawiec PG::InvalidColumnReference: ERROR: for SELECT DISTINCT, ORDER BY expressions must appear in select list. So we probably will have to order in the scope, since the .distinct is there

damianlegawiec commented on core/app/finders/concerns/spree/product_filterable.rb:
```diff
@@ -3,7 +3,7 @@ module ProductFilterable
 
     def find_available(scope, products_scope)
-      scope.filterable.for_products(products_scope)
+      scope.filterable.order("#{OptionType.table_name}.position, #{OptionValue.table_name}.position").for_products(products_scope)
```
@aleksandarpetrushev this concern is also used by properties. You need to move this model-specific code somewhere else

aleksandarpetrushev replied:
@damianlegawiec I just can not get around this error. Even when I select every column. Without .distinct in the :filterable scope of OptionValue it seems to work.

damianlegawiec committed:
Lock active_storage_validations to 0.9.5, 0.9.6 broke our test builds (#11524)

Commit:
Merge branch 'feature/sort-option-values-by-position-in-filter-meta' of https://github.com/spree/spree into feature/sort-option-values-by-position-in-filter-meta

sebastian-palma committed:
Convert added money attributes to string (#11523)

* Convert money display attributes in serializers to string.

* Add missing example cases.

* Fix misspellings.

MatthewKennedy committed:
Fix Sync Taxonomy Name with Root Taxon Name. (#11519)

* Add method for syncing taxonomy name when root taxon name is updated.

* Remove redundant self.

* Use described_class

* reload taxonomy

* Check permalink was changed.

* Check special characters are usage.

* Add check to set root name.

sebastian-palma committed:
[SD-1535] Add the name event to the Webhook serializer payload/body (#11504)

* Add webhooks metadata to webhooks body.

* Create Event after passing validations, before queuing requests.

* Move webhooks metadata to body root.

* Update event instead of create it in HandleRequest.

* Update examples.

* Convert body to json for the request.

* Stub MakeRequest with exact arguments.

* Don't modify body with new events when testing decorators.

* Create Event in HandleRequest with subscriber data.

* Add full flow examples.

* Update event to event_name.

* Convert money display attributes in serializers to string.

* Add missing example cases.

* Fix misspellings.

* Add ActiveJob serializers no error examples.

* Rails 5 can not automatically JSON.dump serializers examples.

MatthewKennedy committed:
Add Custom Reposition Actions - Platform API (#11512)

* Add custom reposition actions.

* Improve reposition action to handle dropping an item in same location it was picked up, without flashing an error.

* Fix spec

* Swaggerize.

* Update child taxon permalinks.

* Add test for taxon permalink rebuild.

Commit:
Order OptionValues by OptionType position then OptionValue position

Commit:
SD-1451 Enable Setting LineItem MetaData when adding Item to Cart (#11525)

* Enable setting metadata when adding item to cart - Storefront API

* Update Add Item to Cart service default params

Co-authored-by: Damian Legawiec <damian@sparksolutions.co>

* Set default arguments to be hashes in add_item service

* Return 422 if Metadata param is not a hash

* Extract metadata param validation to concern

* Rename Metadata Concern and normalize locales file

Co-authored-by: Damian Legawiec <damian@sparksolutions.co>

Commit:
Fix column selecting when sorting OptionValues

Commit:
SD-1471 Geo API - Add States endpoints to Platform API (#11475)

* Add States endpoints to Platform API

* Enable include and filter by country for States API

* Generate Docs

* Fix failing spec for State ransackable attributes

* Update docs

* Update core/app/models/spree/state.rb

* Add name to State ransackable attributes

Co-authored-by: Damian Legawiec <damian@sparksolutions.co>

sebastian-palma committed:
Emit product.out_of_stock. (#11527)

Commit:
Move webhook factories to core/lib (#11531)

* Move webhook factories to core/lib

* Move factories to /lib

damianlegawiec committed:
Stop creating dummy user role (#11533)

This role doesn't have any logic attached and is not needed.

MatthewKennedy committed:
Only return eligible variants (#11534)

* Scope only eligible variants

* Add tests for variants returned on product with or without options.

damianlegawiec committed:
Install mysql gem only when mysql environment is required (#11535)

* Install mysql gem only when mysql environment is required

* fix

Commit:
Add List of custom webhook events in Webhooks::Subscriber (#11495)

* Add List of custom webhook events in Webhooks::Subscriber

* Add Events List to Subscriber model

* Make .defaulte_events public

* Add shipment.shipped event to list of supported custom events

* Generate supported webhook events dynamically

* Fix check for if HasWebhooks is included in model

* Set custom webhook events in decorators

* Rename custom_supported_events to custom_webhook_events

* Remove extra 'end'

* Remove deleted method's specs

* Change event name operation from destroy to delete

* Fix expectation in has_webhooks_spec

Commit:
SD-1452 As a Storefront API user I can set Order MetaData when creating a Cart - clean (#11530)

* Resolve conflicts

* Update Cart#create docs

* Update Create Cart service default params

* Do not nest metadata params in Cart::Create service

* Cast metadata params to hash in Create Cart service

* Whitelist Add Item to Cart params

* Whitelist options as a hash param in CartController

* Move *_metadata params in add_item_params

damianlegawiec committed:
Doorkeeper controllers now inherit from `Api::V2::BaseController` (#11538)

This allows us to call `current_store`, `current_locale` and other Spree methods

damianlegawiec committed:
Removing Variant#track_inventory (#11541)

This method overshadows the `track_inventory` attributes and causes an infinite loop in Rails 7. Both of these reasons are good enough to remove it.

KacperMekarski committed:
failure payment when connection error is raised (#11532)

* failure payment when connection error is raised

* add specs for ActiveMerchant::ConnectionError

Commit:
SD-1544 Do not emit webhooks on touch events (#11539)

* Do not emit webhooks on touch events

* Make check if updating timestamps only be for default events

damianlegawiec committed:
Revert "Only return eligible variants (#11534)" (#11542)

This reverts commit 7e1effe1bb6dc702f07dd2d090d857a223f955a8.

damianlegawiec committed:
Filter out tokens and client id/secret from application logs

damianlegawiec committed:
Ability to create oAuth application with blank redirect_uri

damianlegawiec committed:
Adds encryption for app secrets and access tokens

damianlegawiec committed:
Do not allow access to Platform API for tokens without client associated

This is the default doorkeeper behaviour that we turned off for Storefront API, however we need this as an additional level of protection for Platform API. In Spree v5 we will require also client credentials for Storefront API.

damianlegawiec committed:
Merge pull request #11545 from spree/fix/secure-platform-api

Enhancing Platform API security

damianlegawiec committed:
Fixed test environment for Admin Dashboard (#11546)

Commit:
SD-1544 Check if updated_only_timestamps for update events only (#11544)

* Check if updated_only_timestamps for update events only

* Update has_webhooks.rb

* Use end_with to check if it is an update event

sebastian-palma committed:
Raise exception when event can not be created. (#11536)

mad-eel commented on core/spec/finders/spree/option_values/find_available_spec.rb:
```diff
@@ -68,6 +68,23 @@ module Spree
         end
       end
+
+      context 'ordering' do
```
If I see correctly, there are no cases for sorting by option type position, only option values. At least it's not explicit. Could we expand on it to be sure?

damianlegawiec replied:
<img width="739" alt="Screenshot 2021-12-02 at 10 44 30" src="https://user-images.githubusercontent.com/55154/144397364-ebc92c23-6ba9-4f1c-a093-b317232cb10c.png">

@mad-eel you can sort option types in the Admin, and it should be respected later in the API / UI :)

aleksandarpetrushev replied:
@damianlegawiec I think he means in the specs

Commit:
VEN-99 Run brakeman (#11537)

* brakeman wip

* Setup brakeman config

* Config brakeman in codeclimate.yml

* Force scan with brakeman

* Edit brakeman config

* Brakeman

* Lock brakeman version

* Inline config for brakeman

* Setup brakeman for Spree API - CircleCI

* Run brakeman job

* Fix mistake in circleci config

* Ensure bundle install before running brakeman

* ensure Brakeman installing

* Ensure bundle installed

* Do not exit on warn and error - Brakeman

* Do not run brakeman with bundle

* Up brakeman version

* Setup brakeman for spree/core

* Setup output format and artifacts

* Fix brakeman issues

* Revert secure sql interpolation

* Fail build on brakeman warning

* Ignore some brakeman issues, crash on warning

* Fix brakeman ignore file path

* Change brakeman output format to text

* Test to see if it fails on warning

* Ignore warning

Co-authored-by: Damian Legawiec <damian@getvendo.com>

sebastian-palma committed:
Rename all occurrences of webhooks "body" to "webhook_payload_body". (#11528)

Co-authored-by: Aleksandar Petrushev <aleksandar.petrushev@sparksolutions.co>

Commit:
Sort option values by option type first - specs

Commit:
Sort option values by option type first - specs

Commit:
Remove printing

Commit:
Remove printing

damianlegawiec commented on core/spree_core.gemspec:
```diff
@@ -53,6 +53,6 @@ Gem::Specification.new do |s|
   s.add_dependency 'mini_magick', '~> 4.9', '>= 4.9.4'
   s.add_dependency 'image_processing', '~> 1.2'
-  s.add_dependency 'active_storage_validations', '~> 0.9'
+  s.add_dependency 'active_storage_validations', '~> 0.9', '<= 0.9.5'
```
@aleksandarpetrushev can you rebase this PR with main to remove this change?

Commit:
Resolve conflicts

Commit:
Rebase onto main

Commit:
Resolve conflicts

Commit:
Merge branch 'feature/sort-option-values-by-position-in-filter-meta' of https://github.com/spree/spree into feature/sort-option-values-by-position-in-filter-meta

aleksandarpetrushev commented:
https://github.com/spree/spree/pull/11549 is a new PR for these changes
